### ISSUE TITLE
Missing TS warning in `pipe` method

### ISSUE THREAD

**Issue by fabian-hiller** [fabian-hiller/valibot#669]
The following code should raise a TypeScript warning because it can lead to a runtime error:

```ts
const Schema = v.pipe(v.optional(v.string()), v.transform<string, string>((s) => s.toLowerCase()));
```

**Comment by EltonLobo07** [comment]
This is because method parameter types are bivariant.
The block below shows how the type constraint is created for the trim action.
```
PipeItem<string | null | undefined, unknown, BaseIssue<unknown>>
    transformed to  
PipeAction<string | null | undefined, unknown, BaseIssue<unknown>>
    transformed to
BaseTransformation<string | null | undefined, unknown, BaseIssue<unknown>>
    transformed to
{
    // Other irrelevant keys skipped
    _run(
        dataset: TypedDataset<string | null | undefined, never>,
        // The second parameter is irrelevant for this discussion 
    ): Dataset<unknown, BaseIssue<unknown>>
}
```
The trim action generates an object of the following type:
```
{
    // Other irrelevant keys skipped
    _run(
        dataset: TypedDataset<string, never>,
        // The second parameter is irrelevant for this discussion
    ): Dataset<string, never>
}
```
The comparison boils down to:
Is `TypedDataset<string, never>` assignable to `TypedDataset<string | null | undefined, never>`? 
The answer is yes, because the comparison is between the types of the first parameter of a **method**. Parameter type comparison for a method is bivariant. More information [here](https://github.com/Microsoft/TypeScript/pull/18654).
Also, I have created a simplified example of the root cause [here](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAQghgZwgFQE5wHYIGYHtUC2cwAlrhgDzIB8UAvFAN4BQUbUA+qgK4YAUAE2KIIwAFxMAbnAA23CBOQBfAJQTJuEgOZKA3M2YCIAYxlxU0GaKjBUJAhPhI0mHPiKlyFBLZIYA5tT6Rqbmlta+BADCABYmANYQAgCC-nB+Po4iLlh4hMRklD52AVAAPlAY3DIy5VC8Rth+SUHMVsA2BGD0NnbRccaJKWkZwPoA9ABUrOzIcVCICCT+GAQQGB0ARhAyuADuUCQIlbgdoJBQCHDY0JvcHcgAygs1+8ckWyZw3EhQa8AxXACKBgcxwf4QVA2cAQY5hKCbEjSOyYYAzNgASQ6RxOZxhl2uty+PwU6PYbDm0AABlxeFS-qJAcDyFAqZFYgkkql0lhgPTjJgFsZjBAwB0qcU-P46lUanUGhAmhgklSyeS7md5jSeBh6f8mVAWWy+vTyDIQFB4hg3jZcFAYpgBFZWZKAvTpHJYcxJuNmOyBkNuaMepF9JEQ30OYMuSNefogA).

**Comment by fabian-hiller** [comment]
Thank you for your research and explanation. Do you see a solution to this problem?

**Comment by EltonLobo07** [comment]
Yes, I have a solution that works, but with two minor concerns (I'll talk about the concerns at the end of this reply).

**The solution**:
Create a type from `BaseTransformation` which will be a copy of it but with `_run` as a function than a method and use the created type.
Why? Because it will allow the type checker to use a different kind of comparison ([strict function comparison](https://github.com/Microsoft/TypeScript/pull/18654)), which is exactly what we want for this problem.
Here are the implementation details if we choose to implement it.
1. Create a type from `BaseTransformation` that has a `_run` function instead of a `_run` method.
```ts
export interface BaseTransformationForPipe<
  TInput,
  TOutput,
  TIssue extends BaseIssue<unknown>,
> extends ToFunction<BaseTransformation<TInput, TOutput, TIssue>, "_run"> {

}

type ToFunction<
  T,
  TKey extends keyof T,
> = {
  [K in keyof T]:
    K extends TKey
      ? T[K] extends (...params: infer TParams) => infer TReturn
        ? (...params: TParams) => TReturn
        : T[K]
      : T[K]  
}
```
2. Replace the `BaseTransformation` type with the created type in the union assigned to `PipeAction`.
![PipeAction-type](https://github.com/user-attachments/assets/efd356e8-f901-4cd2-ae76-b5da7d928853)
After step 2, there is a new issue. The `PipeItem<unknown, unknown, BaseIssue<unknown>>` in the `SchemaWithPipe` type won't allow actions that extend the newly created type. This is because the first type argument `unknown` is assigned to `TInput`. The type checker has to check if `TypedDataset<unknown, never>` is assignable to `TypedDataset<string, never>` (for the trim action). Clearly, it is not assignable, so the type checker generates a type error. This is why we need step 3.
3. Convert the type constraints of the `SchemaWithPipe` type parameter (first image) to the type shown in the second image.
![SchemaWithPipe-old](https://github.com/user-attachments/assets/02b37efd-0f95-4e90-b3e2-d565907269d3)
![SchemaWithPipe-new-v2](https://github.com/user-attachments/assets/1bb75203-132c-4190-81b8-16fb6d5c916d)
I think this is safe because the `pipe` function's overloads won't connect mismatched `TOutput` - `TInput` combinations.

**My concerns**:
1. JavaScript assumes the object has a `_run` method instead of a function, but the type checker assumes it's a function. It's not the end of the world, but I don't like this difference.
2. The `any` at step number 3 with the comment to stop Eslint from complaining.

**Final thoughts**:
If you think the type safety the fix will provide is worth the concerns mentioned above and you don't have any other concerns related to the solution, I can work on fixing this issue and raise a PR. Let me know.

**Comment by fabian-hiller** [comment]
Thanks for your research! I am toying with the idea of changing all method declarations to function declarations and also changing all occurrences of `unknown` in generics to `any`. This would lead to stricter and more precise types, but I am not sure if there is any major downside.

**Comment by fabian-hiller** [comment]
The downside of `any seems to be that we lose type safety in our own code. Here is an example: https://github.com/typescript-eslint/typescript-eslint/issues/3038#issuecomment-777671091

**Comment by fabian-hiller** [comment]
I cannot find a workaround for `any`. `unknown` does not work because of the strict function check, and `never` does not work because the generic types are stored as an actual value inside the more specific type and are therefore always defined. [Here](https://www.typescriptlang.org/play/?#code/JYOwLgpgTgZghgYwgAgEJwM4QMoIBYQC2cAPACoB8yA3gFDLID6UAriAFzIAUAbnADYsInMgEpkAXio8A9sAAmAbnrI+g4cjLKAvrVqhIsRCmxgooAOa4CxZBAAekEPIxpMOfEVIYzlqtV1aBBkQH2QfcxArT2JOU0jomzhJGhVmNl4BIXE6BgZg0Jl+CAA6fhkLTPVRZQZtABoVNSFOAHIYGRlWxu1lWhg2BDBgEORIH3JrLztHCGdXdCwp4hI2AGsQGQB3EAoKLgwYuBFluByVKAgwFigQcKOS9JAuds7WmtpA8bAD3yjTj60IA) is a minimal reproduction.

**Comment by EltonLobo07** [comment]
❤️ for the idea of changing all method declarations to function declarations (+ make them readonly).

I apologize for recommending - convert `unknown` type arguments to `any` type. I reaslised I was wrong after reading the linked issue related to `any` type arguments.

In the linked example of your latest comment, the generic type parameter `T` is being used in a covariant position (`value: T`) and also in a contravariant position (`_run: (value: T) => void`) which makes `BaseSchema<T>` invariant.

What if we split the schema into two sub schemas based on the the position of the type parameter and control the generic type parameter with an optional type parameter on the base type as shown [here](https://www.typescriptlang.org/play/?#code/PTAEEEBtNB7AzUAXAFgU1ABwE6027JATwFoA3AQ0gFcNj8BnUAd3WwwAMAVD0ASyaZIFAMZoAJvwB2oCqBGwpSbBUrY+FJVlgM+SPooB0AKD5KC8URgBCFBmgDKI9AFsKAYUXKKANQrYAHi4APlAAb2NQUAB9bGopAC5QAApKGjQkrgBKUABeULJYPnFjAF9jYxAIaDhEVAwcPAJicipaZCJGFjZOHn5BYTFJM1l5WDUNLUwdPQMpQwAC03NsSzFQW3snVw9YP0CQ8MjQNNpMsorlpAsrDbtHZzQ3IIAaGK5PJRV9vNBD47QAA9rlJxAxjlFNg8dp59kFgm8IXcto83J9vHDoh8vN9-KEIhcrjd1g5lGYAObbJ5yIEgsHI6HUgIMMlScn48rGBRSFmgFnqNlUtxJUkCymouS5I5RWLxVJtNA5CJRKLchiwSBoQyQWDk+XpLIAbmOpRex1OGVAAHJ4LBYFazaVjcZ4PERPpFMg0CygkKacC0KCmFC-QF4gBrKSwZhSN5SNBkAjBYLJBgSzJ+pXHNM7QyyqSGlVFqJVegYEihfVnUDxxPYHL5E5FEpRdhIajYGQ56mGC2F0udcuhCNRmOE64s1Os8U7I3GIA)?

**Comment by fabian-hiller** [comment]
I don't know how to proceed at the moment. `never` is not a good option, as it prevents us from calling `._run`, and we do that in many cases. `any` might work, but leads to less type safety internally (less important due to our unit and type tests) and also externally when writing custom functions or integrations (and that's a problem for me). Also, we would force them to use `any` to make it "work". Alternatively, we could investigate and improve your idea and just try to change the types for `pipe`.

I had planned to release our v1 RC in the next few days, but I will postpone it due to this issue, as changes to our base types may result in breaking changes to our ecosystem.

**Comment by EltonLobo07** [comment]
After thinking about it and trying out many different ideas, I think maybe we are trying to capture the wrong types. What if we capture the leaf types instead of types that are higher up? Example [here](https://www.typescriptlang.org/play/?#code/JYOwLgpgTgZghgYwgAgEJwM4QMoIBYQC2cAPACoB8yA3gFDLID6UAriAFzIAUAbnADYsInMgEpkAXio8A9sAAmAbnrI+g4cjLKAvrVqhIsRCmxgooAOa4CxZBAAekEPIxpMOfEVIYzlqtV1aBBkQH2QfcxArT2JOU0jomzhJGhVmNl4BIXE6BgZg0Jl+CAA6fhkLTPVRZQZtABoVNSFOAHIYGRlWxu1lWgB6fuQAUXs4QgAHYs4AYTgJsBYE5GS8YAsCKGQwAE8JiFoYNgQwYBDtiB8ARnJrLztHCGdXdCw74hI2AGsQGQB3EAUChcDAxOAid5wHIqKAQRZQEDhMElZoQHR6QYjMaTabIOYLJaWFbIYpwGDbPYHTHYUBIZAAAzIADUshB6cg-pgVhgMOsQBB5NsZAzvr8AezQNsCA9xlMUHAAEYyHgQerhOEikA-f4gdlcsDS+QQeAsfhgQ7HU7nSA+ABM5BZ6hSop1wNBSU4rw8SQdrIo0No2mQehtYCuIN8UUhNVoodtEYS0eUQA). Maybe this idea can help us avoid the impossible cases we run into right now.

**Comment by fabian-hiller** [comment]
Your idea is great, but I suppose it only works in certain cases. It actually makes sense that we can't pass `stringSchema` to the `test` function, because the test function could pass any `unknown` data to `._run`, but `stringSchema` only takes `string` as the first argument. Failure to pass `string` could result in a runtime error. I think we need to figure out how to change the method declarations to function declarations, but keep `unknown` when using generics. I will try to investigate this tomorrow. I may create a draft PR to share my findings with you.

**Comment by fabian-hiller** [comment]
As I have already written, what TypeScript does makes sense for the most part, and I was able to solve the problem pretty much. Because of the covariant and contravariant thing, we are not able to use `never' in a situation where we don't care about the actual type. Please check my draft PR #785 if you have the time. Even if we are forced to use any sometimes, I think overall the benefits of this change outweigh the drawbacks.

**Comment by EltonLobo07** [comment]
I was able to check the draft PR today. I think the idea is good. To summarize, the `TInput` type parameter of many different types of objects is used as a function parameter type, because of which the `TInput` type arguments of various types were changed from `unknown` to `any`. There are some changes in the draft PR that do not make sense to me. I tried to revert all of the questionable changes, and all of the tests still passed. I have added some questions as comments associated with some of the sections or files that were changed. Can you answer them? I think we should proceed with the idea in the PR.
